src/allocation_engine/interval_tree.rs

```diff
@@ -37,12 +40,93 @@ pub struct Constraint {
 /// memory slot that will be assigned to a device by the VMM.
 /// This structure represents the key of the Node object in
 /// the interval tree implementation.
-#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Hash, Ord)]
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Hash, Ord, Debug)]
 pub struct Range {
     pub min: u64,
     pub max: u64,
 }
 
+impl Range {
+    /// Create a new Range object.
+    pub fn new(min: u64, max: u64) -> Result<Self> {
+        if min > max {
+            return Err(Error::InvalidRange(min, max));
+        }
+        Ok(Range { min: min, max: max })
+    }
+
+    /// Get length of the range.
+    pub fn len(&self) -> u64 {
+        self.max - self.min + 1
+    }
+
+    /// Check whether two Range objects intersect with each other.
+    pub fn intersect(&self, other: &Range) -> bool {
+        max(self.min, other.min) <= min(self.max, other.max)
+    }
+
+    /// Check whether the key is fully covered.
+    pub fn contain(&self, other: &Range) -> bool {
+        self.min <= other.min && self.max >= other.max
+    }
+
+    /// Create a new Range object with min being the first value from the range
```

Comment:
`min` is not very descriptive because it means knowing the implementation details of the structure `Range`. We can replace it with "the start of the region".